keras/src/distribution/tensor_parallel/autoconfig.py

```diff
@@ -0,0 +1,241 @@
+    """
+    if not isinstance(layer, layers.Dense):
+        return "dense"
+
+    input_dim = None
+    output_dim = None
+
+    if hasattr(layer, "_kernel") and layer._kernel is not None:
+        kernel_shape = layer._kernel.shape
+        if len(kernel_shape) == 2:
+            input_dim = kernel_shape[0]
+            output_dim = kernel_shape[1]
+    elif hasattr(layer, "kernel") and layer.kernel is not None:
+        kernel_shape = layer.kernel.shape
+        if len(kernel_shape) == 2:
+            input_dim = kernel_shape[0]
+            output_dim = kernel_shape[1]
+
+    if input_dim is None or output_dim is None:
+        if hasattr(layer, "units"):
+            output_dim = layer.units
+        else:
+            return "dense"
+
+        if (
+            hasattr(layer, "input_shape")
+            and layer.input_shape
+            and len(layer.input_shape) > 1
+        ):
+            input_dim = layer.input_shape[-1]
+        else:
+            return "dense"
+
+    if not input_dim or not output_dim:
+        return "dense"
+
+    expansion_threshold = 1.5
+    is_expansion = output_dim > input_dim * expansion_threshold
+    is_contraction = input_dim > output_dim * expansion_threshold
+
+    if is_expansion:
+        return "up_projection"
+    elif is_contraction:
+        return "down_projection"
+    else:
+        return "dense"
+
+
+def _reduce_sum(x):
+    """Reduces the input tensor across the model axis using sum."""
+    return distribution_lib.all_reduce(x, op="sum", axis_name="model")
+
+
+def _gather(x, axis):
+    """Gathers the input tensor across the model axis along the given axis."""
+    return distribution_lib.all_gather(x, axis=axis, axis_name="model")
+
+
+def _apply_layer_sharding_rules(
+    layer, full_name, device_count, state_rules, output_rules
+):
+    """
+    Applies sharding rules to a single layer instance.
+
+    This function populates the state_rules and output_rules dictionaries with
+    sharding strategies specific to the layer type.
+
+    Args:
+        layer: The Keras layer instance to process.
+        full_name: The full hierarchical name of the layer.
+        device_count: The number of devices available for sharding.
+        state_rules: Dictionary to store parameter sharding rules.
+        output_rules: Dictionary to store output communication rules.
+    """
+
+    def split_rule(dim):
+        return functools.partial(
+            split_tensor_for_parallelism, device_count=device_count, dim=dim
+        )
+
+    def gather_rule(axis):
+        return functools.partial(_gather, axis=axis)
+
+    if isinstance(layer, layers.Dense):
+        mlp_type = analyze_dense_layer(layer)
+
+        if mlp_type == "up_projection":
+            state_rules[f"{full_name}.kernel"] = split_rule(dim=1)
+            if layer.use_bias:
+                state_rules[f"{full_name}.bias"] = split_rule(dim=0)
+            output_rules[f"{full_name}"] = {0: gather_rule(axis=-1)}
```

Comment:
Why `{0: gather_rule(axis=-1)}`? What does the `0` represent?

All of them use `0` always, so why not just do `output_rules[f"{full_name}"] = gather_rule(axis=-1)`?